File: solutions/020_valid_parentheses.go
package main

import "fmt"

// #20 Valid Parentheses [Easy]
// Tags: String, Stack

func isValid(s string) bool {
    dicMap := make(map[string]string,{"(":")","[":"]","{":"}"}

    for i := 1, v := range s {
      before := s[i]

      if v != before

     }
    return false
}

func main() {
	fmt.Println(isValid("()")) // → true
	fmt.Println(isValid("()[]{}")) // → true
	fmt.Println(isValid("(]")) // → false
	fmt.Println(isValid("([])")) // → true
	fmt.Println(isValid("{[]}")) // → true
}

File: solutions/020_valid_parentheses.ts
// #20 Valid Parentheses [Easy]
// Tags: String, Stack
//
// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid. An input string is valid if open brackets are closed by the same type and in the correct order.
//
// Example 1: Input: s = "()"
//          Output: true
// Example 2: Input: s = "()[]{}"
//          Output: true
// Example 3: Input: s = "(]"
//          Output: false

function isValid(s: string): boolean {
  const stack: string[] = [];
  const dic: Record<string,string> = {
    "(":")",
    "{":"}",
    "[":"]"
  } as const;

  for (const ss of s) {

   const toS = dic[ss] ?? "";

   if (toS) {
    stack.push(toS)
   } else {
    const ps = stack.pop()
    if(ss !== ps) {
      return false
    }
   }
  }

  return true;
}

// --- Tests ---
console.log(JSON.stringify(isValid("()"))); // → true
console.log(JSON.stringify(isValid("()[]{}"))); // → true
console.log(JSON.stringify(isValid("(]"))); // → false
console.log(JSON.stringify(isValid("([])"))); // → true
console.log(JSON.stringify(isValid("{[]}"))); // → true
